lib/SILOptimizer/Mandatory/ConstantPropagation.cpp

```diff
@@ -247,6 +247,30 @@ static SILInstruction *constantFoldCompare(BuiltinInst *BI,
     SILBuilderWithScope B(BI);
     return B.createIntegerLiteral(BI->getLoc(), BI->getType(), APInt(1, 1));
   }
+
+  if (match(BI, m_CombineOr(m_BuiltinInst(BuiltinValueKind::ICMP_SLT,
+                                          m_SILValue(Other), m_Zero()),
+                            m_BuiltinInst(BuiltinValueKind::ICMP_SGT, m_Zero(),
+                                          m_SILValue(Other))))) {
+    BuiltinInst *OtherBI = dyn_cast<BuiltinInst>(Other);
+    if (OtherBI &&
+        OtherBI->getBuiltinKind() == BuiltinValueKind::AssumeNonNegative) {
+      SILBuilderWithScope B(BI);
+      return B.createIntegerLiteral(BI->getLoc(), BI->getType(), APInt());
+    }
+  }
+
+  if (match(BI, m_CombineOr(m_BuiltinInst(BuiltinValueKind::ICMP_SGE,
+                                          m_SILValue(Other), m_Zero()),
+                            m_BuiltinInst(BuiltinValueKind::ICMP_SLE, m_Zero(),
+                                          m_SILValue(Other))))) {
+    BuiltinInst *OtherBI = dyn_cast<BuiltinInst>(Other);
+    if (OtherBI &&
+        OtherBI->getBuiltinKind() == BuiltinValueKind::AssumeNonNegative) {
```

Comment:
This can be put into the pattern as:

```
m_CombineOr(
    m_BuiltinInst(BuiltinValueKind::ICMP_SLT,
                          m_BuiltinInst(BuiltinValueKind::AssumeNonNegative, m_ValueBase()),
                          m_Zero()),
    m_BuiltinInst(BuiltinValueKind::ICMP_SGT,
                           m_Zero(),
                           m_BuiltinInst(BuiltinValueKind::AssumeNonNegative, m_ValueBase())) {

}
```